src/Microsoft.AspNet.Mvc.TagHelpers/Internal/AttributeMatcher.cs

```diff
@@ -93,20 +93,26 @@ public static class AttributeMatcher
             var presentAttributes = new List<string>();
             var missingAttributes = new List<string>();
 
-            foreach (var attribute in requiredAttributes)
+            foreach (var requiredAttribute in requiredAttributes)
             {
-                if (!context.AllAttributes.ContainsName(attribute) ||
```

Comment:
My objections are mainly about the unexplained interpretation of `attributeName=""` as a "missing" attribute and (on the flip side) the lack of special handling for minimized attributes. One possible improvement would be to check `TryGetAttribute()` and call it a day. Another would be to find reasons for the other special cases and add comments explaining why they exist.